### PR TITLE
Dump container filesystem instead of extracting layers

### LINKED ISSUE #41: gvltctl build: Layers are not merged - removed files are present in final VM
**NOTE:** This is backport of https://github.com/gevulotnetwork/platform/issues/69.

---

Currently `gvltctl build` unpacks all layers from user image to construct VM filesystem. This works for adding and modifying files, but removed files are not handled! It means that final VM consists of files that were removed between layers.

Example image:

```
FROM alpine:3.20.3

RUN mkdir -p /foo/bar

RUN touch /foo/bar/baz

RUN rm -r /foo
```

Unpacking result:

image

---

In order to handle file removals we would need to handle [whiteouts](https://github.com/opencontainers/image-spec/blob/main/layer.md#whiteouts), but at this point I would rather avoid unpacking individual layers.

### PR DESCRIPTION
**Pull request by koxu1996**
**NOTE:** This is backport of https://github.com/gevulotnetwork/platform/pull/76.

---

Extracting layers should never happen; we want to extract whole filesystem at once.

Fixes #41 and #42.

### CODE REVIEW

File: src/builders/skopeo_builder.rs [hunk 1]
```diff
@@ -310,60 +312,52 @@ impl SkopeoSyslinuxBuilder {
...
         Self::run_command(
             &[
-                "skopeo",
-                "copy",
+                "podman",
+                "create",
+                "--replace",
+                "--name",
+                TEMP_CONTAINER_NAME,
                 container_source,
```
Comment (trusch) [inline src/builders/skopeo_builder.rs:323]: Does this work with all source formats that also skopeo supports? Otherwise we need to document the changes in behavior

Comment (koxu1996) [reply src/builders/skopeo_builder.rs:323]: Good question!

**TL;DR** Yes, all formats are supported.

According to [containers/image docs](https://github.com/containers/image/blob/620f400df1132d1d693a790baae603fcb175e6e2/docs/containers-transports.5.md?plain=1#L11):

> Tools which use the containers/image library, including skopeo(1), buildah(1), podman(1), **all share a common syntax** for referring to container images in various locations.

I double checked that with `man podman create` and the following `transport:path` formats are supported:

- `docker://docker-reference` (**default**)
- `docker-daemon:docker-reference`
- `dir:path`
- `oci-archive:path:tag`
- `docker-archive:path[:docker-reference]`

It almost fully matches with the documented formats in `gvltctl build`:

https://github.com/gevulotnetwork/gvltctl/blob/aad390de567aa6f8a2fbeae34937ad8d8bb5cc30/src/commands/build.rs#L15-L19

The only difference is `docker-daemon` vs `containers-storage`, but they are both local storages. I additionally tested `podman create` with `containers-storage` - it works :wink:.

File: src/builders/skopeo_builder.rs [hunk 2]
```diff
@@ -383,6 +377,25 @@ impl SkopeoSyslinuxBuilder {
...
 
+        // Read image config.
+        let config_path = target_dir.path().join("config.json");
+        let config_path_str = config_path.to_str().unwrap();
+        Self::run_command(
+            &[
+                "sh",
+                "-c",
+                &format!(
+                    "skopeo inspect --config {} > {}",
```
Comment (aleasims) [inline src/builders/skopeo_builder.rs:388]: What about using `podman image inspect` instead of `skopeo inspect` to get rid of skopeo dependency completely?

Looks like because some compatibility issues, `oci_spec::image::ImageConfiguration` and the JSON produced from `podman image inspect` are different (different key cases and etc. with `podman 3.4.4` and `oci_spec 0.7.1`). So this doesn't work for me:

```shell
podman image inspect IMAGE_REF > manifest.json
```

and:

```rust
let manifest = oci_spec::image::ImageConfiguration::from_file("manifest.json")?;
```

But producing only "Config" should work out anyway:

```
podman image inspect IMAGE_REF | jq -rM '.[0].Config' > config.json
```

And now produced config can be deserialized:

```rust
let file = std::io::BufReader::new(std::fs::File::open("config.json")?);
let config: oci_spec::image::Config = serde_json::from_reader(file)?;
```

Alternatively may be just capture the whole JSON and parse it in Rust - probably that would be cleaner and doesn't require `jq`:

```
podman image inspect IMAGE_REF > manifest.json
```

and then:

```rust
let file = std::io::BufReader::new(std::fs::File::open("manifest.json")?);
let manifest_json: serde_json::Value = serde_json::from_reader(file)?;
let config_json = manifest_json.as_array().unwrap()[0].as_object().unwrap()["Config"].clone();
let config: oci_spec::image::Config = serde_json::from_value(config_json)?;
```

Comment (koxu1996) [reply src/builders/skopeo_builder.rs:388]: @aleasims Thanks! I could not figure out how to make it compatible, will try it :crossed_fingers:.

Comment (koxu1996) [reply src/builders/skopeo_builder.rs]: Done in https://github.com/gevulotnetwork/gvltctl/pull/43/commits/d2b9ee61fdb95f3f286608a8330c23475284cb9f, also renamed skopeo builder into podman - https://github.com/gevulotnetwork/gvltctl/pull/43/commits/f494cd03bc2b9331cce3abda7a11c3a9555e6037.

**Comment by koxu1996** [comment]
@trusch @aleasims PTAL :pray:.

**Review by aleasims — APPROVED** [review]
